File: packages/core/heal/loop/loop.go
package loop

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"time"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/handleui/detent/packages/core/heal/tools"
)

const (
	// maxIterations is the internal limit on tool call rounds (not user-configurable).
	maxIterations = 50

	// maxTokensPerResponse is the internal limit on tokens per response.
	maxTokensPerResponse = 8192

	// DefaultTimeout is the total timeout for the healing loop.
	DefaultTimeout = 10 * time.Minute

	// DefaultModel is the model to use for healing.
	DefaultModel = anthropic.ModelClaudeSonnet4_5

	// DefaultBudgetPerRunUSD is the default spending limit per run.
	DefaultBudgetPerRunUSD = 1.00
)

// Config configures the healing loop.
type Config struct {
	Timeout             time.Duration
	Model               anthropic.Model
	BudgetPerRunUSD     float64 // 0 = unlimited
	RemainingMonthlyUSD float64 // -1 = unlimited, 0 = exhausted
	Verbose             bool
}

// DefaultConfig returns a config with default values.
func DefaultConfig() Config {
	return Config{
		Timeout:             DefaultTimeout,
		Model:               DefaultModel,
		BudgetPerRunUSD:     DefaultBudgetPerRunUSD,
		RemainingMonthlyUSD: -1, // unlimited by default
		Verbose:             false,
	}
}

// ConfigFromSettings creates a loop Config from model, timeout, and budget settings.
// This is the canonical way to configure the healing loop from application settings.
// budgetPerRunUSD is the per-run limit (0 = unlimited).
// remainingMonthlyUSD is the remaining monthly budget (-1 = unlimited, 0 = exhausted).
// Verbose mode should be set separately via the Verbose field if needed.
func ConfigFromSettings(model string, timeoutMins int, budgetPerRunUSD, remainingMonthlyUSD float64) Config {
	cfg := DefaultConfig()
	if model != "" {
		cfg.Model = anthropic.Model(model)
	}
	if timeoutMins > 0 {
		cfg.Timeout = time.Duration(timeoutMins) * time.Minute
	}
	if budgetPerRunUSD >= 0 {
		cfg.BudgetPerRunUSD = budgetPerRunUSD
	}
	cfg.RemainingMonthlyUSD = remainingMonthlyUSD
	return cfg
}

// Result contains the outcome of a healing attempt.
type Result struct {
	// Success indicates whether the healing was successful.
	Success bool

	// Iterations is the number of message rounds completed.
	Iterations int

	// FinalMessage is Claude's final response.
	FinalMessage string

	// ToolCalls is the total number of tool calls made.
	ToolCalls int

	// Duration is how long the loop took.
	Duration time.Duration

	// InputTokens is the total input tokens used across all API calls.
	InputTokens int64

	// OutputTokens is the total output tokens used across all API calls.
	OutputTokens int64

	// CacheCreationInputTokens is the total tokens used to create cache entries.
	CacheCreationInputTokens int64

	// CacheReadInputTokens is the total tokens read from cache.
	CacheReadInputTokens int64

	// CostUSD is the calculated cost in USD based on token usage.
	CostUSD float64

	// BudgetExceeded indicates if the loop stopped due to budget limit.
	BudgetExceeded bool

	// BudgetExceededReason indicates which budget was exceeded ("per-run" or "monthly").
	BudgetExceededReason string
}

// HealLoop orchestrates the agentic healing process.
type HealLoop struct {
	client        anthropic.Client
	registry      *tools.Registry
	config        Config
	verboseWriter io.Writer
}

// New creates a new healing loop.
func New(client anthropic.Client, registry *tools.Registry, config Config) *HealLoop {
	var verboseWriter io.Writer
	if config.Verbose {
		verboseWriter = os.Stderr
	}
	return &HealLoop{
		client:        client,
		registry:      registry,
		config:        config,
		verboseWriter: verboseWriter,
	}
}

// Run executes the healing loop with the given system prompt and initial user message.
func (l *HealLoop) Run(ctx context.Context, systemPrompt, userPrompt string) (*Result, error) {
	startTime := time.Now()

	// Apply timeout
	ctx, cancel := context.WithTimeout(ctx, l.config.Timeout)
	defer cancel()

	// Initialize conversation
	messages := []anthropic.MessageParam{
		anthropic.NewUserMessage(anthropic.NewTextBlock(userPrompt)),
	}

	result := &Result{}
	modelName := string(l.config.Model)

	for iteration := range maxIterations {
		result.Iterations = iteration + 1

		// Make API call
		response, err := l.client.Messages.New(ctx, anthropic.MessageNewParams{
			Model:     l.config.Model,
			MaxTokens: maxTokensPerResponse,
			System: []anthropic.TextBlockParam{
				{Text: systemPrompt},
			},
			Messages: messages,
			Tools:    l.registry.ToAnthropicTools(),
		})
		if err != nil {
			result.Duration = time.Since(startTime)
			result.CostUSD = CalculateCostWithCache(modelName, TokenUsage{
				InputTokens:              result.InputTokens,
				OutputTokens:             result.OutputTokens,
				CacheCreationInputTokens: result.CacheCreationInputTokens,
				CacheReadInputTokens:     result.CacheReadInputTokens,
			})
			return result, fmt.Errorf("API call failed: %w", err)
		}

		// Track token usage (including cache tokens)
		result.InputTokens += response.Usage.InputTokens
		result.OutputTokens += response.Usage.OutputTokens
		result.CacheCreationInputTokens += response.Usage.CacheCreationInputTokens
		result.CacheReadInputTokens += response.Usage.CacheReadInputTokens
		result.CostUSD = CalculateCostWithCache(modelName, TokenUsage{
			InputTokens:              result.InputTokens,
			OutputTokens:             result.OutputTokens,
			CacheCreationInputTokens: result.CacheCreationInputTokens,
			CacheReadInputTokens:     result.CacheReadInputTokens,
		})

		// Check per-run budget (don't return error - let caller decide how to handle)
		if l.config.BudgetPerRunUSD > 0 && result.CostUSD > l.config.BudgetPerRunUSD {
			result.BudgetExceeded = true
			result.BudgetExceededReason = "per-run"
			result.Duration = time.Since(startTime)
			return result, nil
		}

		// Check monthly budget (-1 = unlimited, >= 0 = remaining budget)
		if l.config.RemainingMonthlyUSD >= 0 && result.CostUSD > l.config.RemainingMonthlyUSD {
			result.BudgetExceeded = true
			result.BudgetExceededReason = "monthly"
			result.Duration = time.Since(startTime)
			return result, nil
		}

		// Check if we're done (model finished without requesting tools)
		if response.StopReason == anthropic.StopReasonEndTurn {
			result.FinalMessage = l.extractTextContent(response)
			result.Success = true
			result.Duration = time.Since(startTime)
			return result, nil
		}

		// Process tool calls
		var toolResults []anthropic.ContentBlockParamUnion
		hasToolUse := false

		for i := range response.Content {
			block := response.Content[i]
			if toolUse, ok := block.AsAny().(anthropic.ToolUseBlock); ok {
				hasToolUse = true
				result.ToolCalls++

				// Verbose logging
				l.logToolCall(toolUse.Name, toolUse.JSON.Input.Raw())

				// Dispatch the tool
				toolResult := l.registry.Dispatch(ctx, toolUse.Name, json.RawMessage(toolUse.JSON.Input.Raw()))

				// Create tool result block
				toolResults = append(toolResults,
					anthropic.NewToolResultBlock(toolUse.ID, toolResult.Content, toolResult.IsError))
			}
		}

		if !hasToolUse {
			// Model responded with text only, we're done
			result.FinalMessage = l.extractTextContent(response)
			result.Success = true
			result.Duration = time.Since(startTime)
			return result, nil
		}

		// Add assistant response and tool results to conversation
		messages = append(messages,
			response.ToParam(),
			anthropic.NewUserMessage(toolResults...),
		)
	}

	result.Duration = time.Since(startTime)
	return result, fmt.Errorf("max iterations (%d) exceeded", maxIterations)
}

// extractTextContent extracts text content from a response.
func (l *HealLoop) extractTextContent(response *anthropic.Message) string {
	for i := range response.Content {
		if text, ok := response.Content[i].AsAny().(anthropic.TextBlock); ok {
			return text.Text
		}
	}
	return ""
}

// logToolCall logs a tool call in verbose mode with the key parameter.
func (l *HealLoop) logToolCall(toolName, inputRaw string) {
	if l.verboseWriter == nil {
		return
	}

	// Extract key parameter based on tool type
	keyParam := extractKeyParam(toolName, inputRaw)
	if keyParam != "" {
		_, _ = fmt.Fprintf(l.verboseWriter, "  → %s: %s\n", toolName, keyParam)
	} else {
		_, _ = fmt.Fprintf(l.verboseWriter, "  → %s\n", toolName)
	}
}

// keyParamNames maps tool names to their key parameter for verbose output.
var keyParamNames = map[string]string{
	"read_file":   "path",
	"edit_file":   "path",
	"glob":        "pattern",
	"grep":        "pattern",
	"run_check":   "category",
	"run_command": "command",
}

// extractKeyParam extracts the most relevant parameter for verbose output.
func extractKeyParam(toolName, inputRaw string) string {
	var params map[string]any
	if err := json.Unmarshal([]byte(inputRaw), &params); err != nil {
		return ""
	}

	if paramName, ok := keyParamNames[toolName]; ok {
		if value, exists := params[paramName]; exists {
			if s, ok := value.(string); ok {
				// Truncate long values
				if len(s) > 50 {
					return s[:47] + "..."
				}
				return s
			}
		}
	}

	return ""
}

File: packages/healing/src/loop.ts
import type {
  ContentBlock,
  MessageParam,
} from "@anthropic-ai/sdk/resources/messages.js";
import type { Client } from "./client.js";
import { calculateCost } from "./pricing.js";
import type { ToolRegistry } from "./tools/registry.js";
import type { HealConfig, HealResult, TokenUsage } from "./types.js";
import { DEFAULT_CONFIG } from "./types.js";

/**
 * Maximum number of tool call rounds (not user-configurable).
 */
const MAX_ITERATIONS = 50;

/**
 * Maximum tokens per response.
 */
const MAX_TOKENS_PER_RESPONSE = 8192;

/**
 * Maps tool names to their key parameter for verbose output.
 */
const KEY_PARAM_NAMES: Record<string, string> = {
  read_file: "path",
  edit_file: "path",
  glob: "pattern",
  grep: "pattern",
  run_check: "category",
  run_command: "command",
};

/**
 * Extracts the most relevant parameter for verbose output.
 */
const extractKeyParam = (
  toolName: string,
  input: Record<string, unknown>
): string => {
  const paramName = KEY_PARAM_NAMES[toolName];
  if (!paramName) {
    return "";
  }

  const value = input[paramName];
  if (typeof value !== "string") {
    return "";
  }

  if (value.length > 50) {
    return `${value.slice(0, 47)}...`;
  }
  return value;
};

/**
 * Creates the initial result object.
 */
const createInitialResult = (): HealResult => ({
  success: false,
  iterations: 0,
  finalMessage: "",
  toolCalls: 0,
  duration: 0,
  inputTokens: 0,
  outputTokens: 0,
  cacheCreationInputTokens: 0,
  cacheReadInputTokens: 0,
  costUSD: 0,
  budgetExceeded: false,
});

/**
 * Calculates token usage from result.
 */
const getUsageFromResult = (result: HealResult): TokenUsage => ({
  inputTokens: result.inputTokens,
  outputTokens: result.outputTokens,
  cacheCreationInputTokens: result.cacheCreationInputTokens,
  cacheReadInputTokens: result.cacheReadInputTokens,
});

/**
 * Checks if budget limits have been exceeded.
 */
const checkBudgetLimits = (
  config: HealConfig,
  result: HealResult,
  startTime: number
): HealResult | null => {
  if (config.budgetPerRunUSD > 0 && result.costUSD > config.budgetPerRunUSD) {
    return {
      ...result,
      budgetExceeded: true,
      budgetExceededReason: "per-run",
      duration: Date.now() - startTime,
    };
  }

  if (
    config.remainingMonthlyUSD >= 0 &&
    result.costUSD > config.remainingMonthlyUSD
  ) {
    return {
      ...result,
      budgetExceeded: true,
      budgetExceededReason: "monthly",
      duration: Date.now() - startTime,
    };
  }

  return null;
};

/**
 * HealLoop orchestrates the agentic healing process.
 */
export class HealLoop {
  private readonly client: Client;
  private readonly registry: ToolRegistry;
  private readonly config: HealConfig;
  private readonly verboseWriter: ((msg: string) => void) | null;

  constructor(
    client: Client,
    registry: ToolRegistry,
    config: Partial<HealConfig> = {}
  ) {
    this.client = client;
    this.registry = registry;
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.verboseWriter = this.config.verbose
      ? (msg: string) => process.stderr.write(msg)
      : null;
  }

  /**
   * Executes the healing loop with the given system prompt and initial user message.
   */
  run = async (
    systemPrompt: string,
    userPrompt: string
  ): Promise<HealResult> => {
    const startTime = Date.now();
    const messages: MessageParam[] = [{ role: "user", content: userPrompt }];
    const result = createInitialResult();
    const api = this.client.api;
    const modelName = this.config.model;

    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(
        () => reject(new Error("Healing loop timeout exceeded")),
        this.config.timeout
      );
    });

    try {
      for (let iteration = 0; iteration < MAX_ITERATIONS; iteration++) {
        result.iterations = iteration + 1;

        const response = await Promise.race([
          api.messages.create({
            model: modelName,
            max_tokens: MAX_TOKENS_PER_RESPONSE,
            system: systemPrompt,
            messages,
            tools: this.registry.toAnthropicTools(),
          }),
          timeoutPromise,
        ]);

        this.updateTokenUsage(result, response.usage, modelName);

        const budgetExceeded = checkBudgetLimits(
          this.config,
          result,
          startTime
        );
        if (budgetExceeded) {
          return budgetExceeded;
        }

        if (response.stop_reason === "end_turn") {
          return this.createSuccessResult(result, response.content, startTime);
        }

        const { toolResults, hasToolUse } = await this.processToolCalls(
          response.content,
          result
        );

        if (!hasToolUse) {
          return this.createSuccessResult(result, response.content, startTime);
        }

        messages.push(
          { role: "assistant", content: response.content },
          { role: "user", content: toolResults }
        );
      }

      result.duration = Date.now() - startTime;
      result.finalMessage = `Max iterations (${MAX_ITERATIONS}) exceeded`;
      return result;
    } catch (error) {
      result.duration = Date.now() - startTime;
      result.costUSD = calculateCost(modelName, getUsageFromResult(result));
      result.finalMessage =
        error instanceof Error ? error.message : "Unknown error occurred";
      return result;
    }
  };

  /**
   * Updates the result with token usage from response.
   */
  private readonly updateTokenUsage = (
    result: HealResult,
    usage: {
      input_tokens: number;
      output_tokens: number;
      cache_creation_input_tokens?: number | null;
      cache_read_input_tokens?: number | null;
    },
    modelName: string
  ): void => {
    result.inputTokens += usage.input_tokens;
    result.outputTokens += usage.output_tokens;
    result.cacheCreationInputTokens += usage.cache_creation_input_tokens ?? 0;
    result.cacheReadInputTokens += usage.cache_read_input_tokens ?? 0;
    result.costUSD = calculateCost(modelName, getUsageFromResult(result));
  };

  /**
   * Creates a success result with final message.
   */
  private readonly createSuccessResult = (
    result: HealResult,
    content: ContentBlock[],
    startTime: number
  ): HealResult => ({
    ...result,
    finalMessage: this.extractTextContent(content),
    success: true,
    duration: Date.now() - startTime,
  });

  /**
   * Processes tool calls from response content.
   */
  private readonly processToolCalls = async (
    content: ContentBlock[],
    result: HealResult
  ): Promise<{
    toolResults: Array<{
      type: "tool_result";
      tool_use_id: string;
      content: string;
      is_error?: boolean;
    }>;
    hasToolUse: boolean;
  }> => {
    const toolResults: Array<{
      type: "tool_result";
      tool_use_id: string;
      content: string;
      is_error?: boolean;
    }> = [];
    let hasToolUse = false;

    for (const block of content) {
      if (block.type === "tool_use") {
        hasToolUse = true;
        result.toolCalls++;

        const input = block.input as Record<string, unknown>;
        this.logToolCall(block.name, input);

        const toolResult = await this.registry.dispatch(block.name, input);
        toolResults.push({
          type: "tool_result",
          tool_use_id: block.id,
          content: toolResult.content,
          is_error: toolResult.isError || undefined,
        });
      }
    }

    return { toolResults, hasToolUse };
  };

  /**
   * Extracts text content from a response.
   */
  private readonly extractTextContent = (content: ContentBlock[]): string => {
    for (const block of content) {
      if (block.type === "text") {
        return block.text;
      }
    }
    return "";
  };

  /**
   * Logs a tool call in verbose mode with the key parameter.
   */
  private readonly logToolCall = (
    toolName: string,
    input: Record<string, unknown>
  ): void => {
    if (!this.verboseWriter) {
      return;
    }

    const keyParam = extractKeyParam(toolName, input);
    if (keyParam) {
      this.verboseWriter(`  -> ${toolName}: ${keyParam}\n`);
    } else {
      this.verboseWriter(`  -> ${toolName}\n`);
    }
  };
}

/**
 * Creates a config from settings.
 * This is the canonical way to configure the healing loop from application settings.
 */
export const createConfig = (
  model: string,
  timeoutMins: number,
  budgetPerRunUSD: number,
  remainingMonthlyUSD: number
): HealConfig => ({
  timeout: timeoutMins > 0 ? timeoutMins * 60_000 : DEFAULT_CONFIG.timeout,
  model: model || DEFAULT_CONFIG.model,
  budgetPerRunUSD:
    budgetPerRunUSD >= 0 ? budgetPerRunUSD : DEFAULT_CONFIG.budgetPerRunUSD,
  remainingMonthlyUSD,
  verbose: false,
});
